[14.0][NEW] l10n_br_bank_api_inter

pablomatoskmee opened the pull request:
Adição do modulo l10n_br_bank_api_inter atualizada para Odoo 14.0.

pablomatoskmee committed:
[NEW][14.0]l10n_br_bank_api_inter

[IMP] adds fields for auth in bank api

[FIX] set correct name for model account.move

[FIX] Workflow for validate and consult pdf of the invoice for Inter

[FIX] Print all invoices in account.move model

[FIX] Allows to search the invoice stats

ygcarvalh commented on l10n_br_bank_api_inter/demo/account_payment_order.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Se as funções desse arquivo não vão ser utilizadas, pode remover ele.

ygcarvalh commented on l10n_br_bank_api_inter/models/account_invoice.py:
```diff
@@ -0,0 +1,95 @@
+            }
+        except Exception as error:
+            raise UserError(error)
+            # raise error
```
Pode remover essa linha

ygcarvalh commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,216 @@
+    bank_inter_state = fields.Selection(
+        selection=ESTADO,
+        string="State",
+        # readonly=True,
```
Remover essa linha

ygcarvalh commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,216 @@
+
+        order_id = self.payment_line_ids[0].order_id
+        with ArquivoCertificado(order_id.journal_id, "w") as (key, cert):
+            # self.journal_id.bank_account_id
```
Remover essa linha

ygcarvalh commented on l10n_br_bank_api_inter/tests/test_l10n_br_bank_api_inter.py:
```diff
@@ -0,0 +1,308 @@
+        with self.assertRaises(UserError):
+            payment_order.open2generated()
+
+    # TODO: Esse caso de teste está com problema, pois com a validação de cancelar a
```
Você pode corrigir esse teste?

pablomatoskmee replied:
Das solicitações somente esse teste eu não fui capaz de encontrar uma solução.

ygcarvalh commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,216 @@
+                with ArquivoCertificado(order.order_id.journal_id, "w") as (key, cert):
+                    self.api = ApiInter(
+                        cert=(cert, key),
+                        conta_corrente=(
```
Onde possível, coloque o nome das variáveis em inglês, isso ajuda a ter uma padronização no código

ygcarvalh commented on l10n_br_bank_api_inter/models/account_payment_order.py:
```diff
@@ -0,0 +1,122 @@
+                issue_date=line.create_date,
+                due_date=line.date,
+                identifier=line.name,
+                instructions=[
```
Está certo essas instruções de testes?

ygcarvalh commented on l10n_br_bank_api_inter/models/account_payment_order.py:
```diff
@@ -0,0 +1,122 @@
+                )
+                if payment_line_id:
+                    payment_line_id.move_line_id.own_number = resposta["nossoNumero"]
+                # bank_line_id.seu_numero = resposta['seuNumero']
```
Remover essa linha se não for utilizar

ygcarvalh commented on l10n_br_bank_api_inter/parser/inter_file_parser.py:
```diff
@@ -0,0 +1,502 @@
+        # e caso se queira saber os detalhes será preciso olhar a Entrada de
+        # Diário referente.
+
+        if self._code_log() == data["situacao"]:
```
Essa condição não pode ser um problema pela função não estar implementada?

ygcarvalh commented on l10n_br_bank_api_inter/parser/inter_file_parser.py:
```diff
@@ -0,0 +1,502 @@
+        if self._code_log() == data["situacao"]:
+            return
+
+        valor_titulo = data["valorNominal"]
```
Onde possível, coloque o nome das variáveis em inglês

ygcarvalh commented:
@pablomatoskmee dá uma olhada no erro do runboat também

AndreMarcos commented:
@pablomatoskmee, se puder dar uma olhada novamente no runboat novamente

mileo commented on l10n_br_bank_api_inter/models/account_invoice.py:
```diff
@@ -0,0 +1,94 @@
+    def action_invoice_cancel(self):
+        try:
+            financial_move_line_ids = self.financial_move_line_ids
+            if financial_move_line_ids.bank_inter_state == "baixado":
```
```
        to_cancel = self.financial_move_line_ids.filtered(lambda fml: fml.bank_inter_state == "baixado")
        if to_cancel:
            to_cancel.drop_bank_slip()
```
Sugestão para corrigir o erro do runbot.

marcelsavegnago commented:
@pablomatoskmee consegue aumentar a cobertura de teste ?

pablomatoskmee commented:
> 

@marcelsavegnago 
Sim, To trabalhando no aumento dessa cobertura.

rvalyi commented:
seria bom atingir uns 90% de cobertura. Hoje temos um pouco mais de 87% na v14 mas a gente gostaria ter 90% na v16, o que nem seria algo excepcional na OCA. Se vc olhar os desenvolvedores mudaram muito na Kmee, 3 ou 4 anos depois quando quem desenvolveu não participa mais, é essencial ter esses testes senão acaba sendo bucha para as outras pessoas que mantém o projeto.

ygcarvalh committed:
[ADD] Set readme files for the module

Commit:
[ADD] Creates a wizard for the user to select the reason for the dismissal.

[ADD] Consulta Automática de Boleto

B

[ADD]Consulta automática boleto

[FIX] Corrige o manifest

GCKrauss committed:
[ADD] Scheduled Action Test

GCKrauss committed:
[ADD] Schedule action to check the status of bill (boleto) daily

mbcosta commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,237 @@
+
+_logger = logging.getLogger(__name__)
+
+try:
```
É desnecessário importar a biblioteca com o TRY quando a lib é informada no manifest no atributo external_dependencies isso já é feito, mais detalhes https://github.com/OCA/l10n-brazil/pull/2744

mbcosta commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,237 @@
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    pdf_boleto_id = fields.Many2one(
```
Existe necessidade de armazenar esse arquivo no account.move.line ? Foi algo especifico ou algum caso de uso que levou a isso? Porque pelo o que vejo os Boletos acabam sendo salvos em apenas um arquivo e se necessário o re-envio é mandando esse arquivo com todas as Parcelas e não um arquivo com uma única Parcela, ou vários arquivos cada um para uma Parcela

mbcosta commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,237 @@
+                "target": "new",
+            }
+
+    def drop_bank_slip_wizard(self):
```
Não seria melhor ao invés de criar um novo método buscar integração com o que já existe no l10n_br_account_payment_order o "def create_cnab_write_off" https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_account_payment_order/models/l10n_br_cnab_change_methods.py#L307 ? Ou teve algum motivo que não permitiu

mbcosta commented on l10n_br_bank_api_inter/models/account_payment_order.py:
```diff
@@ -0,0 +1,120 @@
+    _logger.error("Biblioteca erpbrasil.bank.inter não instalada")
+
+try:
+    from febraban.cnab240.user import User, UserAddress, UserBank
```
Parece que faltou incluir essa lib no manifest

mbcosta commented on l10n_br_bank_api_inter/models/account_payment_order.py:
```diff
@@ -0,0 +1,120 @@
+    _logger.error("Biblioteca febraban não instalada")
+
+try:
+    from erpbrasil.base import misc
```
Parece que faltou incluir essa lib no manifest

mbcosta commented on l10n_br_bank_api_inter/models/account_move.py:
```diff
@@ -0,0 +1,94 @@
+
+    _inherit = "account.move"
+
+    pdf_boletos_id = fields.Many2one(
```
Tanto o campo para armazenar o arquivo PDF do Boleto como método para abrir foram incluídos no l10n_br_account_payment_order aqui https://github.com/OCA/l10n-brazil/pull/2724 para evitar duplicidade de código e quando mais de um modulo referentes a uma API ou CNAB de Banco estiverem implementados não ter diversos botões exemplo "Imprimir Boleto API Inter", "Imprimir Boleto API XY" e etc, por isso é melhor buscar a integração com o modulo

mbcosta commented on l10n_br_bank_api_inter/models/account_payment_order.py:
```diff
@@ -0,0 +1,120 @@
+
+    def generate_payment_file(self):
+        self.ensure_one()
+        try:
```
É desnecessário fazer esse TRY em todo método, o melhor e fazer apenas no "return self._generate_bank_inter_api()" linha 116, os outros casos devem ser tratados pelo super

mbcosta commented on l10n_br_bank_api_inter/parser/inter_file_parser.py:
```diff
@@ -0,0 +1,502 @@
+        #     lambda b: b.mov_instruction_code_id.id
+        #               == payment_line.payment_mode_id.cnab_sending_code_id.id
+        # )
+        bank_line = payment_line.bank_line_id
```
Esse campo foi removido, referencia https://github.com/OCA/l10n-brazil/pull/2386

mbcosta commented on l10n_br_bank_api_inter/parser/inter_file_parser.py:
```diff
@@ -0,0 +1,502 @@
+        #   necessário que seja implementado os outros métodos suportados pelo Banco
+        #   Inter principalmente na visão do Odoo, já que ele considera apenas uma
+        #   forma de desconto e o Banco Inter aceita até três, cada uma com uma data
+        #   diferente de validade.
```
É possível implementar esses campos de Desconto nesse modulo, vai precisar ser visto de isolar esse caso e se existe a necessidade de criar outras Contas Contábeis

mbcosta commented on l10n_br_bank_api_inter/wizard/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Arquivos init que tem apenas import estão sendo incluídos sem o cabeçalho de licença

mbcosta commented on l10n_br_bank_api_inter/models/account_move_line.py:
```diff
@@ -0,0 +1,237 @@
+
+                        if resposta["situacao"].lower() != self.bank_inter_state:
+                            if resposta["situacao"] == "pago":
+                                move_id = self.env["account.move"].create(
```
Aqui não deveria estar sendo usado o que foi feito arquivo do parser https://github.com/kmee/l10n-brazil/blob/14.0-l10n_br_bank_api_inter/l10n_br_bank_api_inter/parser/inter_file_parser.py#L442 ? Outro ponto o Odoo não permite reconciliar Valores A Menor ou A Maior do que o valor Original então aqui não está faltando lançar os valores de Juros, Multa e Desconto?

mbcosta commented:
Estou revendo a implementação do l10n_br_account_payment_order antes da migração, se houver algo que vocês acreditam ser necessário ou que pode ser alterado para ter uma maior compatibilidade ou diminuir código duplicado com esse modulo isso pode ser visto, nos testes recomendo incluir os casos de Valor A Menor e A Maior para validar essa questão, já que isso é algo comum.

rvalyi reviewed (changes requested):
Teria que ver as observações do @mbcosta dar um rebase e tb botar um test coverage de uns 90% para não afundar as metricas de qualidade do projeto.

marcelsavegnago commented:
@mileo consegue dar um feedback sobre as questoes levantadas ? Um rebase já estaria valendo.
